CONTRIBUTING: remove suggestion about the order of arguments (#247424)
NixOS/nixpkgs: doc/contributing/coding-conventions.chapter.md

  As an exception, an explicit conditional expression with null can be used when fixing a important bug without triggering a mass rebuild.
  If this is done a follow up pull request _should_ be created to change the code to `lib.optional(s)`.

- The order of elements in the argument set should follow the convention below:
  1. `lib`;
  2. `stdenv`;
  3. builders (e.g. `buildPythonPackage`, `mkDerivation`, `buildGoModule`);
  4. fetchers (e.g. `fetchurl`, `fetchpatch`, `fetchFromGitHub`);
  5. all the remaining arguments (i.e. those that don't fit the other items);
  6. options (e.g. `extraModules`);
  7. Boolean flags (e.g. `enableSDL`);

>   8. `buildPackages`;

Author: I will put this before the other items, to be sure, between 4 and 5.

  9. Darwin-related arguments (e.g. `darwin`, `IOKit`, `Metal`).

>   Each item is optional, in the sense it should not be included when not needed (e.g. if there is no need for `buildPackages`, then it should not be included).

Reviewer: ```suggestion
  Each item is optional, in the sense it should not be included when not needed. For example, if there is no need for `buildPackages`, then it should not be included.
```

Avoid nesting, it makes parsing harder.

>   Inside each item of the hierarchy above, the arguments should be in "ASCIIbetical" order, (e.g. `SDL` comes before `Xaw3d`, that comes before `alsa-lib`, that comes before `assemblyscript`).

Reviewer: ```suggestion
  Inside each item of the hierarchy above, the arguments should be in "ASCIIbetical" order.
Example:`SDL`, `Xaw3d`, `alsa-lib`, `assemblyscript`.
```

Just make a concrete example, avoid filling words that don't convey information - that it easier to scan. 

Is there a particular reason for this oddly specific convention though?

  As an exception, an explicit conditional expression with null can be used when fixing a important bug without triggering a mass rebuild.
  If this is done a follow up pull request _should_ be created to change the code to `lib.optional(s)`.

- The order of elements in the argument set must follow the convention below:

  01. `lib`;

>   02. `stdenv`;

Reviewer 2: I would argue that this should be merged into the builders, if not, other types of stdenv should be mentioned as well, e.g. `clangStdenv`, `stdenvNoCC`

Author: "`stdenv` and friends" would be a good name?

Or you are also suggesting to include/lump stdenvs and builders here?

Reviewer 2: > Or you are also suggesting to include/lump stdenvs and builders here?

I would prefer that, yes

>   03. builders (`buildPythonPackage`, `mkDerivation`, `buildGoModule`);

Reviewer 2: "builders" needs to be defined here, does `rustPlatform` count? what about `makeRustPlatform`? what about `python3` if you are just using `python3.pkgs.buildPythonApplication`?

Author: > "builders" needs to be defined here, does `rustPlatform` count? what about `makeRustPlatform`?

Now I am a bit confused. Is `rustPlatform` something similar to `stdenv`?
There are other `languagePlatforms` similar to it, so I can compare?

For now I am inclined to include them here too.

> what about `python3` if you are just using `python3.pkgs.buildPythonApplication`?

This is still `python3`. So, item 6.

I believe this is consistent because you are not seeking for a typical builder, but by the Python-ecosystem-provided one.

Reviewer 2: `rustPlatform` is essentially a wrapper for `cargo`, `rustc`, `cargo-auditable`, and `stdenv`, and exposes things like `buildRustPackage` (a builder, which the most commonly directly used), `fetchCargoTarball` (fod fetcher that takes a `src`), and `rustLibSrc` (derivation)

Author: Basically a stdenv with rusteroids.
Fine. Let's lump them.

  04. fetchers (`fetchurl`, `fetchpatch`, `fetchFromGitHub`);

>   05. `buildPackages`;

Reviewer 2: other sets should also be mentioned here, e.g. `pkgsBuildBuild` and `pkgs`

Author: Nice.

  06. all the remaining arguments (i.e. those that don't fit the other items);
  07. options (`extraModules`, `disabledProtocols ? [ "tcp"  "udp" ]`);
  08. Boolean flags (`enableSDL`);

>   09. testers (`nixosTests`, `testers`);

Reviewer 2: A common pattern in tests is

```nix
testers.testVersion {
  package = <attribute>;
}
```

Since a lot of builders still doesn't support the `finalAttrs` pattern, `<attribute>` would need to be added to the inputs, maybe attributes like this should be added to this section as well?

  10. updaters (`gitUpdater`, `gitUnstableUpdater`);

>   11. Darwin-related arguments (`darwin`, `Metal`, `IOKit`).

Reviewer 2: I don't think darwin should be special cased here, this can be merged into `all the remaining arguments`

Author: This is more because of tradition. They always appeared last, sometimes with a comment like `# Darwin dependencies`.

I believe this is good, because they are dealt only when some specific Darwin quirk comes into the picture.

Keeping Darwin dependencies in a dedicated block helps when someone proposes a Darwin-specific change.
